File: Рабочий стол/source/proxy-gateway-master/src/gpg/serializers/task.py
from rest_framework import serializers

from gpg.serializers import CheckPointSerializer
from gpg.serializers.comment_task import CommentTaskSerializer
from gpg.serializers.sub_task import SubTaskSerializer
from gpg.serializers.task_file import TaskFileSerializer
from gpg.serializers.task_nested_previous_task import TaskNestedPreviousTaskSerializer
from gpg.services import TaskRepresentationService


class TaskSerializer(serializers.Serializer):
    id = serializers.UUIDField(read_only=True)
    name = serializers.CharField(max_length=255)
    description = serializers.CharField(max_length=2555)
    executor = serializers.CharField(max_length=255)
    check_point = CheckPointSerializer(required=False, allow_null=True)
    reviewers = serializers.ListField(required=False)
    responsible = serializers.ListField(required=False)
    observers = serializers.ListField(required=False)
    members = serializers.ListField(required=False)
    is_selected_reviewer = serializers.BooleanField(required=False)
    is_task_agreement = serializers.BooleanField(required=False)
    priority = serializers.ChoiceField(choices=[0, 1, 2])
    status = serializers.ChoiceField(choices=[0, 1, 2, 3, 4, 5, 6])
    type = serializers.ChoiceField(choices=[0, 1, 2, 3])
    previous_task = TaskNestedPreviousTaskSerializer(required=False, allow_null=True)
    next_task = TaskNestedPreviousTaskSerializer(required=False, allow_null=True)
    project = serializers.CharField(max_length=255)
    progress = serializers.FloatField(read_only=True)
    started_at = serializers.DateTimeField()
    planned_ended_at = serializers.DateTimeField()
    ended_at = serializers.DateTimeField(required=False)
    updated_by = serializers.CharField(max_length=55, required=False)
    created_by = serializers.CharField(max_length=55, required=False)
    updated_at = serializers.DateTimeField(read_only=True)
    created_at = serializers.DateTimeField(read_only=True)
    sub_tasks = SubTaskSerializer(many=True)
    task_files = TaskFileSerializer(many=True)
    task_comments = CommentTaskSerializer(many=True)

    def to_representation(self, instance):
        representation = super().to_representation(instance)
        # representation['checkpoint'] =
        representation = TaskRepresentationService(representation)()

        return representation

    def create(self, validated_data):
        raise NotImplementedError()

    def update(self, instance, validated_data):
        raise NotImplementedError()

File: Рабочий стол/source/spa-dev/src/boot/dto/task.js
export const prepareTaskToCreate = (taskData) => {
  return {
    name: taskData.name,
    description: taskData.description,
    priority: taskData.priority,
    project: taskData?.project?.name,
    check_point: taskData?.check_point?.id,
    previous_task: taskData?.previous_task?.id,
    next_task: taskData?.next_task?.id,
    members: taskData?.members.map(member => member.id),
    is_selected_reviewer: taskData?.selected_reviewer,
    is_task_agreement: taskData?.task_agreement,
    reviewers: taskData?.reviewers.map(reviewer => reviewer.id),
    responsible: taskData?.responsible.map(reviewer => reviewer.id),
    executor: taskData.executor.id,
    observers: taskData?.observers.map(observer => observer.id),
    status: taskData.status,
    type: taskData.type,
    started_at: taskData.started_at,
    planned_ended_at: taskData.planned_ended_at,
    sub_tasks: taskData.sub_tasks.map(subtask => {
      return {
        name: subtask.name,
        is_completed: subtask.is_completed,
        executor: subtask.executor.id,
      }
    }),
  }
}

export const prepareTaskToUpdate = (oldTask, taskData) => {
  const newTask = {
    id: taskData.id
  }

  for (const [key, value] of Object.entries(taskData)) {
    if (value !== oldTask[key]) {
      if (['responsible', 'members', 'observers', 'reviewers'].some(i => key.includes(i))) {
        newTask[key] = value.map(user => user.id)
      } else if (['check_point', 'executor', 'previous_task', 'next_task'].some(i => key.includes(i))) {
        newTask[key] = value.id
      } else if (['started_at', 'planned_ended_at'].some(i => key.includes(i))) {
        if (new Date(value).toISOString().replace(/\:\d{2}\.\d+Z$/, '') !== new Date(oldTask[key]).toISOString().replace(/\:\d{2}\.\d+Z$/, '')) {
          newTask[key] = value
        }
      } else if (['project'].some(i => key.includes(i))) {
        newTask[key] = value.name
      } else {
        newTask[key] = value
      }
    }
  }


  delete newTask.task_files
  delete newTask.updated_at
  delete newTask.updated_by
  delete newTask.sub_tasks
  delete newTask.created_at
  delete newTask.created_by
  delete newTask.task_comments


  return newTask
}
